Disentagle gossip/snitch dependencies

avikivity opened the pull request:
Gossip and snitch have a circular dependency: snitch depends on
gossip to disseminate rack/dc info, while gossip talks to snitch to
ask its name and to tell it it is starting. The circular dependency
is fixed up at runtime by having separate construct and start times,
so it sort of works, but relies on global variables for snitch and 
gossiper that we'd like to remove.

This series breaks the circular dependency by moving the gossiper
parts of the dependency to the callers - both the snitch name and
the gossiper starting notification are now provided by the caller.

This paves the way for the elimination of gossiper and snitch global
variables later.

Test: unit (dev)

avikivity commented:
@vladzcloudius @asias review ping

asias commented:
> @vladzcloudius @asias review ping

Will do.

vladzcloudius commented on main.cc:
```diff
@@ -836,6 +836,7 @@ int main(int ac, char** av) {
 
             gms::gossip_config gcfg;
             gcfg.gossip_scheduling_group = dbcfg.gossip_scheduling_group;
+            gcfg.snitch_name = locator::i_endpoint_snitch::get_local_snitch_ptr()->get_name();
```
This is not invariant and can be changed in a runtime via REST API: https://github.com/scylladb/scylla/blob/master/api/storage_service.cc#L686

vladzcloudius replied:
I don't see how this is API is any useful. However we may want to either take care of that or remove this API in the context of this change.

xemul replied:
The only place where gossiper calls snitch for its name is the `check_snitch_matches()`, which in turn is only called on start. So in order for the API to wreak havoc is to get triggerred until the node finishes its shadow round. Is that really _that_ awful?

vladzcloudius replied:
For the sake of correctness AFAICT check_snitch_name_matches() is called from two places: storage_service::check_for_endpoint_collision(...) and storage_service::prepare_to_join(...).

But what if this changes in the future? My point was about a "spaghetti" code we are creating here. If we want to make a Snitch name to be an invariant - let's make it such and I see no issue whatsoever to drop that API which nobody uses as far as I'm aware of.

xemul replied:
If not dropping the API, let's then add the `snitch_name` argument, all the more so its peer `check_knows_remote_features` accepts checked features as arguments, instead of getting them itself.

xemul replied:
@vladzcloudius I'm all for removing the API, no code is always better :) But if not, then -- the `check_for_endpoint_collision` is called from storage service, my proposal is to get snitch name in those places, (storage service already gets it [nearby](https://github.com/scylladb/scylla/blob/c38ceafdcfca87b3706b55fe9c73a7f015a4b560/service/storage_service.cc#L349)) and pass it as argument into `check_for_endpoint_collision`.

vladzcloudius replied:
Yup. As long as we don't cache it - we should be safe enough.

avikivity commented:
v2:
 - rebased
 - removed the restriction that the snitch cannot be updated (we'll update the name if the api call happens)
 - retested

I agree we should remove the ability to update the snitch altogether, I just don't want to do it here.

vladzcloudius commented on gms/gossiper.cc:
```diff
@@ -1860,10 +1860,7 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
 
-        return replicate(get_broadcast_address(), local_state).then([] {
-            //notify snitches that Gossiper is about to start
-            return locator::i_endpoint_snitch::get_local_snitch_ptr()->gossiper_starting();
```
One of the important things that is going to happen in the context of this call is a registration of a `reconnectable_helper` for snitches like GossipingPropertyFileSnitch and Ec2MultiDcSnitch. 

This callback is meant to enforce usage of 'local' addresses for inter-node RPC between nodes in the same DC.

In the original code it is registered before, for instance, a `_failure_detector_loop_done` assignment. And there are a few other `gossiper` state variables set here after the original call for a `gossiper_starting()`.

With the new code the `reconnect_helper` is going to be registered after all this.

@asias Could you, please, confirm that it's not going to break anything?

avikivity replied:
@asias ping

asias replied:
Will do.

asias replied:
Btw. @avikivity  Let's delete test/manual/gossip.cc. It is not useful now. 

```

The triplication of the call is unfortunate. One duplication can be
removed by deleting test/manual/gossip.cc, which likely hasn't been
run in several years. The other duplication needs careful study of
storage_service startup.
```

asias replied:
@avikivity I think you missed a place  to call 

locator::i_endpoint_snitch::get_local_snitch_ptr()->gossiper_starting().get()

in the bottom of storage_service::prepare_to_join where 

_gossiper.start_gossiping(generation_number, app_states, advertise).get();

is called.

Current patch does break things ;(

It is sad we need to call gossiper_starting everywhere. It is very easy to get it wrong. 

If you do want to reduce the dependency, why not just wrap the snitch start into a function object and pass it to gossiper.

vladzcloudius reviewed (commented):
Need Asias to confirm that the reordering is safe for production snitches.

xemul commented on api/storage_service.cc:
```diff
@@ -773,10 +773,15 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
-        return locator::i_endpoint_snitch::reset_snitch(ep_snitch_class_name).then([] {
-            return make_ready_future<json::json_return_type>(json_void());
+        return locator::i_endpoint_snitch::reset_snitch(ep_snitch_class_name).then([&g] {
+            auto name = locator::i_endpoint_snitch::get_local_snitch_ptr()->get_name();
```
Let's capture the `ep_snitch_class_name`. The `get_local_snitch_ptr()` is another global that's going to be eliminated anyway.

avikivity replied:
ep_snitch_class_name can change by reser_snitch

xemul commented on service/storage_service.cc:
```diff
@@ -1783,10 +1783,8 @@ future<> storage_service::start_gossiping() {
-                ss._gossiper.start_gossiping(utils::get_generation_number()).then([&ss] {
-                    ss._initialized = true;
-                }).get();
-                undo.cancel();
```
The `undo` thing is orphanized by this patch.

avikivity commented:
@xemul your work probably obsoleted this, no?

xemul commented:
@avikivity , no, this PR is still relevant. I'd be happy to merge it after rebase and CI pass

avikivity commented:
First two patches are obsolete.

The last patch is impossible - it adds a dependency gossip -> get snitch name, but now snitch has gossip as a dependency. So we need to introduce a layer of indirection.

xemul commented:
> but now snitch has gossip as a dependency.

`[PATCH v2 00/23] Maintain dc/rack by topology` removes it

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2719/

xemul commented:
```
[2022-10-18T15:41:31.616Z] [1/2438]   alternator release [ FAIL ] run.1 0.99s
...
[2022-10-18T15:43:31.309Z] Failed to start server at host 127.169.200.1.
[2022-10-18T15:43:31.309Z] Check the log files:
[2022-10-18T15:43:31.309Z] /jenkins/workspace/releng/Scylla-CI/scylla/testlog/x86_64/test.py.release.log
[2022-10-18T15:43:31.309Z] /jenkins/workspace/releng/Scylla-CI/scylla/testlog/x86_64/release/scylla-1.log
```

and scylla-1.log

```
INFO  2022-10-18 18:43:30,474 [shard 0] init - creating tracing
INFO  2022-10-18 18:43:30,474 [shard 0] init - starting API server
INFO  2022-10-18 18:43:30,474 [shard 0] init - Scylla API server listening on 127.169.200.1:10000 ...
INFO  2022-10-18 18:43:30,475 [shard 0] service_level_controller - update_from_distributed_data: starting configuration polling loop
INFO  2022-10-18 18:43:30,475 [shard 0] init - starting system keyspace
INFO  2022-10-18 18:43:30,475 [shard 0] init - starting gossiper
INFO  2022-10-18 18:43:30,475 [shard 0] init - seeds={127.169.200.1}, listen_address=127.169.200.1, broadcast_address=127.169.200.1
scylla: ./seastar/include/seastar/core/sharded.hh:796: Service &seastar::sharded<locator::snitch_ptr>::local() [Service = locator::snitch_ptr]: Assertion `local_is_initialized()' failed.
Aborting on shard 0.
Backtrace:
  0x4dece38
  0x4e1c5e2
  0x7f2a10a50a1f
  /lib64/libc.so.6+0x3d2a1
  /lib64/libc.so.6+0x268a3
  /lib64/libc.so.6+0x26788
  /lib64/libc.so.6+0x35a15
  0x10d3ad1
  0x10c2c15
  0x10c2c05
  0x10c2bf5
  0x10c2ad3
  0x506b411
```

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2726/
